t/protocol_directnic.pl

```diff
@@ -0,0 +1,231 @@
+        wantreq => '/dns/gateway/abc/?data=2001:db8::1',
+        wantstatus => {
+            h1 => {'status-ipv6' => 'good', 'ipv6' => '2001:db8::1', 'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+    {
+        desc => 'IPv4 and IPv6, single host, good',
+        cfg => {h1 => {urlv4 => "$hostname/dns/gateway/abc/", urlv6 => "$hostname/dns/gateway/def/"}},
+        wantips => {h1 => {wantipv4 => '192.0.2.1', wantipv6 => '2001:db8::1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'good', 'ipv4' => '192.0.2.1',
+                'status-ipv6' => 'good', 'ipv6' => '2001:db8::1',
+                'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv4/},
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+    {
+        desc => 'IPv4 and IPv6, single host, mixed success',
+        cfg => {h1 => {urlv4 => "$hostname/dns/gateway/bad_token/", urlv6 => "$hostname/dns/gateway/def/"}},
+        wantips => {h1 => {wantipv4 => '192.0.2.1', wantipv6 => '2001:db8::1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'failed',
+                'status-ipv6' => 'good', 'ipv6' => '2001:db8::1',
+                'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/There was an error updating your record/},
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+);
+
+for my $tc (@test_cases) {
+    diag('==============================================================================');
+    diag("Starting test: $tc->{desc}");
+    diag('==============================================================================');
+    local $ddclient::globals{debug} = 1;
+    local $ddclient::globals{verbose} = 1;
+    my $l = Logger->new($ddclient::_l);
+    local %ddclient::config;
+    $ddclient::config{$_} = {
+        protocol => 'directnic',
+        %{$tc->{cfg}{$_}},
+    } for keys(%{$tc->{cfg}});
+    # Run init_config(), this is what loads JSON support based on the protocols defined in the config
+    ddclient::init_config();
+    # init_config() will filter out config properties, but the part of update_nics() that populates wantip*
+    # is hard to extract, so we'll just wire in our desired wantip* config directly here
+    $ddclient::config{$_} = {
+        %{$ddclient::config{$_}},
+        %{$tc->{wantips}{$_}},
+    } for keys(%{$tc->{wantips}});
+    {
+        local $ddclient::_l = $l;
+        ddclient::nic_directnic_update(sort(keys(%{$tc->{cfg}})));
+    }
+    # We check only the keys of $tc->{wantstatus} against the properties in %ddclient::config, to avoid having to
+    # check all the undef keys that are initialized by init_config()
```

Comment:
You should also check that certain keys are _not_ set (such as `wantipv4` and friends).  That's why `t/protocol_dyndns2.pl` explicitly lists the keys to check: https://github.com/ddclient/ddclient/blob/13a66c79bb90becd183b691d186748d9ebc63ed4/t/protocol_dyndns2.pl#L270-L279